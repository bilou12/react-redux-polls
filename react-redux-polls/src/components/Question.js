import { connect } from "react-redux";
import { userHasAlreadyVoted } from "../utils/helper";
import { handleSaveQuestionAnswer } from "../actions/shared";
import LoginPage from "./LoginPage";

const Question = ({ dispatch, question, usersValues, authedUser }) => {
  if (authedUser === null || authedUser === "") {
    return <LoginPage />;
  }

  let questionAuthor = question.author;

  let user = usersValues.filter((u) => {
    return u.id === questionAuthor;
  })[0];

  let canVote = !userHasAlreadyVoted(question, authedUser);

  const handleOnClick = (e) => {
    e.preventDefault();
    dispatch(handleSaveQuestionAnswer(authedUser, question.id, e.target.value));
  };

  return (
    <div className="question2">
      <div>
        <p>{"Poll by: " + user.name}</p>
        <br></br>
        <img
          src={user.avatarURL}
          alt={`Avatar of ${user.id}`}
          className="avatar"
        />
      </div>
      <div>
        <p>Would you rather?</p>
      </div>
      <div>
        {canVote && (
          <div>
            <p>{question.optionOne.text}</p>
            <br></br>
            <span>or</span>
            <br></br>
            <p>{question.optionTwo.text}</p>
            <button onClick={handleOnClick} value="optionOne">
              Option 1
            </button>
            <button onClick={handleOnClick} value="optionTwo">
              Option 2
            </button>
          </div>
        )}
      </div>
      <div>
        {!canVote && (
          <div>
            <p>{question.optionOne.text}</p>
            <p>
              {question.optionOne.votes.length} votes =
              {Math.round(
                (100 * question.optionOne.votes.length) /
                  (question.optionOne.votes.length +
                    question.optionTwo.votes.length)
              )}
              %
            </p>
            {question.optionOne.votes.includes(authedUser) && (
              <div>
                <p>You voted for the option one above</p>
              </div>
            )}
            <br></br>
            <p>{question.optionTwo.text}</p>
            <p>
              {question.optionTwo.votes.length} votes =
              {Math.round(
                (100 * question.optionTwo.votes.length) /
                  (question.optionOne.votes.length +
                    question.optionTwo.votes.length)
              )}
              %
            </p>
            {question.optionTwo.votes.includes(authedUser) && (
              <div>
                <p>You voted for the option two above</p>
              </div>
            )}
          </div>
        )}
      </div>
    </div>
  );
};

const mapStateToProps = ({ users, authedUser }) => {
  return {
    usersValues: Object.values(users),
    authedUser,
  };
};

export default connect(mapStateToProps)(Question);
