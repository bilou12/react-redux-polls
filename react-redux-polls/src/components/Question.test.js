import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import Question from "./Question";

const mockStore = configureMockStore();
const store = mockStore({
  users: {
    user1: {
      id: "user1",
      name: "User 1",
      avatarURL: "avatar1.png",
    },
    user2: {
      id: "user2",
      name: "User 2",
      avatarURL: "avatar2.png",
    },
  },
  authedUser: "user1",
});

describe("Question component", () => {
  it("should match snapshot", () => {
    const question = {
      id: "question1",
      author: "user1",
      optionOne: {
        text: "Option 1",
        votes: ["user1"],
      },
      optionTwo: {
        text: "Option 2",
        votes: [],
      },
    };

    const { container } = render(
      <Provider store={store}>
        <Question question={question} />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
