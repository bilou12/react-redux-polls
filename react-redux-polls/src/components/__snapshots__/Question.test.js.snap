// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question component should match snapshot 1`] = `
<div>
  <div
    class="question2"
  >
    <div>
      <p>
        Poll by: User 1
      </p>
      <br />
      <img
        alt="Avatar of user1"
        class="avatar"
        src="avatar1.png"
      />
    </div>
    <div>
      <p>
        Would you rather?
      </p>
    </div>
    <div />
    <div>
      <div>
        <p>
          Option 1
        </p>
        <p>
          1
           votes =
          100
          %
        </p>
        <div>
          <p>
            You voted for the option one above
          </p>
        </div>
        <br />
        <p>
          Option 2
        </p>
        <p>
          0
           votes =
          0
          %
        </p>
      </div>
    </div>
  </div>
</div>
`;
